src/app/modules/submission/encrypt-submission/IncomingEncryptSubmission.class.ts

```diff
@@ -0,0 +1,109 @@
+import { Result } from 'neverthrow'
+
+import { FieldIdSet, getVisibleFieldIds } from '../../../../shared/util/logic'
+import { FieldResponse, IPopulatedEncryptedForm } from '../../../../types'
+import { checkIsEncryptedEncoding } from '../../../utils/encryption'
+import { IncomingSubmission } from '../IncomingSubmission.class'
+import {
+  ConflictError,
+  ProcessingError,
+  ValidateFieldError,
+} from '../submission.errors'
+import {
+  ValidatedFieldMap,
+  VerifiableResponseIdSet,
+  VisibleResponseIdSet,
+} from '../submission.types'
+import { getFilteredResponses } from '../submission.utils'
+
+export default class IncomingEncryptSubmission extends IncomingSubmission {
+  private constructor(
+    responses: FieldResponse[],
+    public readonly form: IPopulatedEncryptedForm,
+    public readonly encryptedContent: string,
+  ) {
+    super(responses, form)
+  }
+
+  static init(
+    form: IPopulatedEncryptedForm,
+    responses: FieldResponse[],
+    encryptedContent: string,
+  ): Result<
+    IncomingEncryptSubmission,
+    ProcessingError | ConflictError | ValidateFieldError
+  > {
+    return checkIsEncryptedEncoding(encryptedContent)
+      .andThen(() => this.processResponses(responses, form))
+      .map(
+        (filteredResponses) =>
+          new IncomingEncryptSubmission(
+            filteredResponses,
+            form,
+            encryptedContent,
+          ),
+      )
+  }
+
+  private static processResponses(
+    responses: FieldResponse[],
+    form: IPopulatedEncryptedForm,
+  ): Result<
+    FieldResponse[],
+    ProcessingError | ConflictError | ValidateFieldError
+  > {
+    return getFilteredResponses(form, responses)
+      .andThen((filteredResponses) =>
+        this.getFieldMap(form, filteredResponses).map((fieldMap) =>
+          this.transformFilteredResponses({
+            form,
+            filteredResponses,
+            fieldMap,
+          }),
+        ),
+      )
+      .andThen((responsesAndMetadata) =>
+        this.validateResponses({ ...responsesAndMetadata, form }).map(
+          () => responsesAndMetadata.responses,
+        ),
+      )
+  }
+
+  private static transformFilteredResponses({
+    form,
+    filteredResponses,
+    fieldMap,
+  }: {
+    form: IPopulatedEncryptedForm
+    filteredResponses: FieldResponse[]
+    fieldMap: ValidatedFieldMap
+  }): {
+    responses: FieldResponse[]
+    fieldMap: ValidatedFieldMap
+    visibleFieldIds: FieldIdSet
+    verifiableResponseIds: VerifiableResponseIdSet
+    visibleResponseIds: VisibleResponseIdSet
```

Comment:
help me understand something - why take the trouble to pass large amounts of data between static functions when we can just have things like `visibleFieldIds` and `verifiableResponseIds` as class members, and declare functions like `transformFilteredResponses` as class members too? that way you get an encapsulated class which can perform processing and validation internally without having to keep track of so many params

Reply:
we dont want 1) constructors to throw, and 2) instantiated objects to hold invalid data. This means that we need a separate validation (in this case `init`) method that does the validation for us before instantiating. The problem here is that although it these metadata will be useful as instance fields (and will be saved as such in `IncomingEmailSubmission`), we can't rely on them during the `init` step since instantiation can't happen yet

Reply:
(1) won't happen anyway due to neverthrow, and doesn't having a private constructor address (2)? `init` can create an object and call methods on it to validate itself, and return an error if the validation fails. so any consumer of the class can only receive valid objects.

imo the advantage of having all these data as class members is that we can call methods on these objects in future.

Reply:
constructors cant return neverthrow objects, so sadly we still need to rely on an external validation method prior instantiation.

When i picture the flow of data in a scenario that uses these class members for validation, the less-than-ideal events i see are

1) Potentially invalid `IncomingSubmission` instances circulate in memory: Even if we eventually discard and return an error, ideally we don't instantiate them in the first place. i draw similarities to saving an invalid submission to the DB first, and then deleting if it fails validation.

2) Metadata that is used purely during validation (for now) ends up getting tracked for the lifetime of the instance.

I do I agree with the advantage you mentioned though. When we add features to the storage mode submission pipeline these metadata could come in handy.

Given that these metadata will eventually be saved in `IncomingEmailSubmission`, I can always just shift it up to the parent `IncomingSubmission` class so both child classes track metadata.

Let me revise the `init` pipeline to use the instantiate-then-validate pattern and revisit this discussion. maybe the code cleanliness benefit might just outweigh everything

Reply:
>constructors cant return neverthrow objects, so sadly we still need to rely on an external validation method prior instantiation.

yeah, I think what I meant was that if the constructor is just meant to instantiate class members (eg `verifiableFieldIds`, `visibleFieldIds`etc) then it shouldn't throw anyway

Reply:
>Potentially invalid IncomingSubmission instances circulate in memory

I'm not sure this is how software works leh. it's not like a virus where things can circulate in an uncontrolled manner right? imo very different from saving invalid documents to the db, where the invalid state is clearly persisted, unlike having an object in memory where it's very clear when it goes out of scope.

Reply:
it's not a 100% accurate analogy, but it's also a non-ideal behaviour especially if in the future, devs build on the init pipeline and incorrectly assume that all IncomingEncryptSubmission objects are pre-validated

Reply:
i made changes to the init pipeline, lmk what you think!